src/coord/src/coord.rs

```diff
@@ -822,6 +848,42 @@ impl<S: Append + 'static> Coordinator<S> {
             .unwrap();
     }
 
+    async fn group_commit(&mut self) {
+        if !self.pending_writes.is_empty() {
```

Comment:
This check happens on the main Coordinator thread, which can cause unnecessary busy work for the main coord thread. Currently a background thread unconditionally sends a `GroupCommit` message to the main thread every 1 ms. An alternative would be to wrap `pending_writes` in a `Arc<Mutex<_>>`, then have the background thread only send a message if `pending_writes` is not empty.

Reply:
I think the current approach works fine, I just wanted to highlight a potential perf trade-off in case anyone had strong feelings.

Reply:
Oh, yeah, this seems super unfortunate. I'd recommend restructuring so that when you add a pending write it works like this:

```
let pending_write = /* ... */;

if self.pending_writes.is_empty() {
    let cmd_tx = self.internal_cmd_tx.clone();
    tokio::spawn(async {
        tokio::time::sleep(Duration::from_millis(1)).await;
        cmd_tx.send(Message::GroupCommit);
    });
}

self.pending_writes.push(pending_write);
```

If you wanted to avoid the task spawn, you could introduce a field `group_commit_at: Instant` on the coordinator, and then adjust the main `select!` in `serve` like so:

```rust
let msg = select! {
    // Order matters here. We want to process internal commands
    // before processing external commands.
    biased;

    Some(m) = internal_cmd_rx.recv() => m,
    m = self.dataflow_client.ready() => {
        let () = m.unwrap();
        Message::ControllerReady
    }
    m = cmd_rx.recv() => match m {
        None => break,
        Some(m) => Message::Command(m),
    },
    _ = tokio::time::sleep_until(self.group_commit_at) if !self.pending_writes.is_empty() => {
        Message::GroupCommit
    }
};
```

and then similarly you'd set `group_commit_at` to be `Instant::now() + Duration::from_millis(1))` if you were the first write in `self.pending_writes`.

Reply:
I like the first approach suggested. Every transaction already spawns a new task that handles blocking the client until the transaction is complete, so we can just piggy back on that. I just pushed an update that adds the 1ms wait to that task.